test_integration/02_stake_and_mint/02_stake_and_mint_proof.js

```diff
@@ -0,0 +1,329 @@
+
+const StakeAssertion = require('./utils/stake_assertion');
+const ProveGatewayAssertion = require('../lib/prove_gateway_assertion');
+const ConfirmStakeIntentAssertion = require('./utils/confirm_stake_intent_assertion');
+const ProgressStakeAssertion = require('./utils/progress_stake_assertion');
+const ProgressMintAssertion = require('./utils/progress_mint_assertion');
+const ProofUtils = require('../lib/proof_utils');
+const Anchor = require('../lib/anchor');
+
+
+let assertStake;
+let assertProgressStake;
+let assertProgressMint;
+let proofUtils;
+
+let originAccounts;
+let auxiliaryAccounts;
+let originWeb3;
+let auxiliaryWeb3;
+
+let gateway;
+let cogateway;
+let token;
+let baseToken;
+let originAnchor;
+let auxiliaryAnchor;
+let ostPrime;
+let stakeRequest;
+
+const MessageStatus = {
+  Undeclared: 0,
+  Declared: 1,
+  Progressed: 2,
+  DeclaredRevocation: 3,
+  Revoked: 4,
+};
+
+describe('Stake and mint (with proof)', async () => {
+  beforeEach(async () => {
+    originWeb3 = shared.origin.web3;
+    auxiliaryWeb3 = shared.auxiliary.web3;
+    originAccounts = shared.origin.accounts;
+    auxiliaryAccounts = shared.origin.accounts;
+    token = shared.origin.contracts.Token;
+    baseToken = shared.origin.contracts.BaseToken;
+    gateway = shared.origin.contracts.EIP20Gateway;
+    cogateway = shared.auxiliary.contracts.EIP20CoGateway;
+    ostPrime = shared.auxiliary.contracts.OSTPrime;
+
+    const hasher = Utils.generateHashLock();
+    stakeRequest = {
+      amount: new BN(200),
+      gasPrice: new BN(1),
+      gasLimit: new BN(100),
+      staker: originAccounts[0],
+      bounty: await gateway.bounty.call(),
+      nonce: await gateway.getNonce.call(originAccounts[0]),
+      beneficiary: auxiliaryAccounts[2],
+      hashLock: hasher.l,
+      unlockSecret: hasher.s,
+    };
+
+    assertStake = new StakeAssertion(gateway, token, baseToken);
+    assertProgressStake = new ProgressStakeAssertion(gateway, token, baseToken);
+    assertProgressMint = new ProgressMintAssertion(auxiliaryWeb3, cogateway, ostPrime);
+    proofUtils = new ProofUtils(originWeb3, auxiliaryWeb3);
+    originAnchor = new Anchor(
+      auxiliaryWeb3,
+      shared.origin.contracts.Anchor,
+      shared.origin.organizationAddress,
+    );
+    auxiliaryAnchor = new Anchor(
+      originWeb3,
+      shared.auxiliary.contracts.Anchor,
+      shared.auxiliary.organizationAddress,
+    );
+  });
+
+  it('Stake & Mint (progress stake and then progress mint)', async () => {
+    await stake();
+    await confirmStakeIntent();
+    await progressStakeWithProof(MessageStatus.Declared);
+    await progressMintWithProof(MessageStatus.Progressed);
+  });
+
+  it('Stake & Mint (progress mint and then progress stake)', async () => {
+    await stake();
+    await confirmStakeIntent();
+    await progressMintWithProof(MessageStatus.Declared);
+    await progressStakeWithProof(MessageStatus.Progressed);
```

Comment:
Progressing mint before stake (with proof)? 😕

Reply:
That is possible.

Reply:
I thought you need to prove that stake was progressed? I am confused ☹️